File: 2021년/7월/0712 Two Dots.cpp
#include <iostream>

using namespace std;

int n,m,visited[50][50];
int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};
char graph[50][50];

void print()
{
	cout << "Yes";
	exit(0);
}

void dfs(int x, int y, int first_x, int first_y, int cnt)
{
	if (cnt >= 4)
	{
		for(int i=0;i<4;i++)
		{
			if(first_x == x + dx[i] and first_y == y + dy[i]) print();
		}
		
	}
}

int main()
{
	cin >> n >> m;
	for(int i=0;i<n;i++)
	{
		for(int j=0;j<m;j++) cin >> graph[i][j];
	}
	
	for(int i=0;i<n;i++)
	{
		for(int j=0;j<m;j++)
		{
			if(visited[i][j] == 0)
			{
				visited[i][j] = 1;
				dfs(i,j,i,j,0);
			}
		}
	}
	
}

File: 2021년/7월/0712 Two Dots.py
# 백준_16929_Two Dots_DFS_골드 4

import sys
input = sys.stdin.readline

def dfs(now_x,now_y,cnt):
    global start_x, start_y

    if cnt >= 4:
        for i in range(4):
            if now_x + dx[i] == start_x and now_y + dy[i] == start_y:
                print("Yes")
                exit(0)

    for i in range(4):
        next_x = now_x + dx[i]
        next_y = now_y + dy[i]
        if 0<=next_x<n and 0<=next_y<m and visited[next_x][next_y] == 0 and graph[next_x][next_y] == graph[start_x][start_y]:
            visited[next_x][next_y] = 1
            dfs(next_x,next_y,cnt+1)


n,m = map(int,input().split())
graph = []
visited = [[0]*m for _ in range(n)]
dx = [1,-1,0,0]
dy = [0,0,1,-1]

for _ in range(n):
    graph.append(list(input()))

for i in range(n):
    for j in range(m):
        if visited[i][j] == 0:
            start_x = i
            start_y = j

            visited = [[0]*m for _ in range(n)]
            visited[i][j] = 1
            dfs(i,j,1)

print("No")